src/operations/blas3/gemm_partial_local.hpp

```diff
@@ -0,0 +1,454 @@
+  static constexpr bool trans_a = TransA;
+  static constexpr bool trans_b = TransB;
+
+  /* Work groups per dimension m, n, k */
+  const index_t group_count_m;
+  const index_t group_count_n;
+  const index_t group_count_k;
+
+  /* The number of tiles to be processed */
+  const index_t num_tiles;
+
+  SYCL_BLAS_INLINE GemmPartial(input_t A, input_t B, output_t cube_buffer,
+                               element_t alpha, index_t wg_count_k)
+      : a_(A),
+        b_(B),
+        cube_(cube_buffer),
+        alpha(alpha),
+        m_(a_.get_size_row()),
+        n_(b_.get_size_col()),
+        k_(a_.get_size_col()),
+        lda_(a_.getSizeL()),
+        ldb_(b_.getSizeL()),
+        ldc_(cube_.getSizeL()),
+        group_count_m((m_ - 1) / tile_size_dim_m + 1),
+        group_count_n((n_ - 1) / tile_size_dim_n + 1),
+        group_count_k(wg_count_k),
+        num_tiles((k_ - 1) / (tile_size_dim_k * group_count_k) + 1) {}
+
+  void bind(cl::sycl::handler& h) {
+    a_.bind(h);
+    b_.bind(h);
+    cube_.bind(h);
+  }
+  void adjust_access_displacement() {
+    a_.adjust_access_displacement();
+    b_.adjust_access_displacement();
+    cube_.adjust_access_displacement();
+  }
+
+  /*!
+   * @brief This function returns the depth of the cube buffer that should give
+   * the best performance.
+   */
+  static SYCL_BLAS_INLINE index_t get_ideal_cube_depth(index_t compute_units,
+                                                       index_t m, index_t n,
+                                                       index_t k) noexcept {
+    const index_t group_count_mn =
+        ((m - 1) / tile_size_dim_m + 1) * ((n - 1) / tile_size_dim_n + 1);
+    /* The depth of the cube buffer is calculated so that each compute unit
+     * will compute 4 work groups. This value is empirical */
+    return (4 * compute_units - 1) / group_count_mn + 1;
+  }
+
+  /*!
+   * @brief This function is used to find the optimum number of work groups
+   * required to execute each partial GEMM.
+   */
+  SYCL_BLAS_INLINE index_t
+  get_workgroup_cluster(index_t compute_units) noexcept {
+    return ((m_ - 1) / tile_size_dim_m + 1) * ((n_ - 1) / tile_size_dim_n + 1) *
+           group_count_k;
+  }
+
+  /*!
+   * @brief Get the nd_range value which has to be used for kernels that
+   * intend to call GemmPartial::run().
+   */
+  SYCL_BLAS_INLINE cl::sycl::nd_range<1> get_nd_range(
+      index_t compute_units) noexcept {
+    const cl::sycl::range<1> nwg(get_workgroup_cluster(compute_units));
+    const cl::sycl::range<1> wgs(local_thread_size);
+    return cl::sycl::nd_range<1>(nwg * wgs, wgs);
+  }
+
+  template <typename local_memory_t>
+  SYCL_BLAS_INLINE void eval(local_memory_t scratch,
+                             cl::sycl::nd_item<1> id) noexcept {
+    /* Pointers to the scratch memory (lhs and rhs) */
+    auto scratch_ptr = scratch.localAcc.get_pointer().get();
+    auto rhs_scratch_ptr = scratch_ptr + (2 * ld_lhs_tile * tile_size_dim_k);
```

Comment:
just make double buffering optional. (DoubleBuffer +1)*ld_lhs_tile*tile_size_dim_k

Reply:
Done, tested and benchmarked (spoiler: it's less performant if you disable double buffering)